### PR TITLE
Allow test cancellations

### PR DESCRIPTION
**Pull request by kargakis**
Fixes https://github.com/kubernetes/test-infra/issues/3788

### CODE REVIEW

**Comment by BenTheElder** [comment]
image
:thinking: :thinking:

**Comment by kargakis** [comment]
Yeah, different git client :)

File: prow/jenkins/jenkins.go [hunk 1]
```diff
@@ -292,3 +294,21 @@ func (c *Client) GetLog(job string, build int) ([]byte, error) {
 	}
 	return buf, nil
 }
+
+// Abort a Jenkins build that is either running or waiting in the queue.
+func (c *Client) Abort(job string, build *JenkinsBuild) error {
+	u := fmt.Sprintf("%s/job/%s/%s/stop", c.baseURL, job, build.Number)
```
Comment (cjwagner) [inline prow/jenkins/jenkins.go]: `build.Number` should use the `%d` format specifier instead of `%s`.

File: prow/jenkins/controller.go [hunk 2]
```diff
@@ -220,13 +221,19 @@ func (c *Controller) terminateDupes(pjs []kube.ProwJob) error {
 			toCancel = prev
 			dupes[n] = pj
 		}
+		if build, exists := jbs[pj.Metadata.Name]; exists {
```
Comment (cjwagner) [inline prow/jenkins/controller.go]: You need to use `toCancel` in this section instead of `pj`.

File: prow/jenkins/jenkins.go [hunk 3]
```diff
@@ -292,3 +297,29 @@ func (c *Client) GetLog(job string, build int) ([]byte, error) {
...
+	u := fmt.Sprintf("%s/job/%s/%d/stop", c.baseURL, job, build.Number)
+	var header map[string]string
+	if build.IsEnqueued() {
+		// Openshift CI runs a Jenkins proxy that uses a local cache of job names
+		// to proxy requests to multiple Jenkins masters. In case there is a queued
+		// build that needs to be cancelled, the proxy has no idea what is the
+		// correct master to redirect to because the queue cancellation request bears
+		// no information about the job for which the build needs to be cancelled.
+		// The custom header is used by the proxy to figure the correct master out.
+		header = map[string]string{"Jenkins-Job": job}
```
Comment (kargakis) [inline prow/jenkins/jenkins.go]: @stevekuznetsov I don't like this at all. How about using a `max_concurrency` in the operator that will enables us to stay away from filling the queue? Cancelling superseded running jobs works like a charm.

File: prow/jenkins/controller.go [hunk 4]
```diff
@@ -221,6 +222,16 @@ func (c *Controller) terminateDupes(pjs []kube.ProwJob) error {
 			dupes[n] = i
 		}
 		toCancel := pjs[cancelIndex]
+
+		// Allow cancelling superseded PR jobs.
```
Comment (stevekuznetsov) [inline prow/jenkins/controller.go]: not a useful comment

File: prow/jenkins/jenkins.go (hunk 3 above)
Comment (stevekuznetsov) [reply prow/jenkins/jenkins.go]: Not sure I follow the comment -- isn't the `POST` containing the job name and build number?

File: prow/jenkins/jenkins.go [hunk 5]
```diff
@@ -292,3 +297,29 @@ func (c *Client) GetLog(job string, build int) ([]byte, error) {
...
+		u = fmt.Sprintf("%s/queue/cancelItem?id=%d", c.baseURL, build.ID)
+	}
+	resp, err := c.request(http.MethodPost, u, header)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		return fmt.Errorf("response not 2XX: %s: (%s)", resp.Status, u)
```
Comment (stevekuznetsov) [inline prow/jenkins/jenkins.go]: What other than 200 is acceptable?

File: prow/jenkins/jenkins.go (hunk 3 above)
Comment (kargakis) [reply prow/jenkins/jenkins.go]: Discussed in IRC, we can filter these prowjobs out and not terminate any duplicates as long as they are in the queue.

File: prow/jenkins/jenkins.go (hunk 5 above)
Comment (stevekuznetsov) [reply prow/jenkins/jenkins.go]: OK, may want to log non-200 in case there is some weird response we end up chasing in the future

Comment (kargakis) [reply prow/jenkins/jenkins.go]: Yeah, this error is logged in the caller.

**Comment by kargakis** [comment]
@cjwagner @stevekuznetsov this is ready for a last review.

File: prow/jenkins/controller.go [hunk 6]
```diff
@@ -221,6 +222,21 @@ func (c *Controller) terminateDupes(pjs []kube.ProwJob) error {
 			dupes[n] = i
 		}
 		toCancel := pjs[cancelIndex]
+		// Allow aborting presubmit jobs for commits that have been superseded by
+		// newer commits in Github pull requests.
+		if c.ca.Config().JenkinsOperator.AllowCancellations {
+			build, buildExists := jbs[toCancel.Metadata.Name]
+			// Avoid cancelling enqueued builds.
+			if buildExists && build.IsEnqueued() {
+				continue
```
Comment (stevekuznetsov) [inline prow/jenkins/controller.go]: explanatory comment _why_ might be useful here

**Comment by stevekuznetsov** [comment]
@spxtr @cjwagner what added approval here?

**Comment by cjwagner** [comment]
> @spxtr @cjwagner what added approval here?

#4870

**Comment by stevekuznetsov** [comment]
dope

**Pull request merged**